server/src/main/java/org/elasticsearch/common/filesystem/LinuxFileSystemNatives.java

```diff
@@ -0,0 +1,230 @@
+                    throw ue2;
+                }
+            }
+        }
+        logger.debug("library {} is registered", method.getSimpleName());
+        this.method = method;
+    }
+
+    static LinuxFileSystemNatives getInstance() {
+        return INSTANCE;
+    }
+
+    private abstract static class BaseMethod {
+
+        /**
+         * char *strerror(int errnum)
+         *
+         * https://linux.die.net/man/3/strerror
+         */
+        private static native String strerror(int errno);
+    }
+
+    private static final class Stat64Method extends BaseMethod {
+
+        private Stat64Method() {}
+
+        /**
+         * stat(const char *path, struct stat *buf)
+         *
+         * https://linux.die.net/man/2/stat64
+         */
+        private static native int stat64(String path, Stat64 stats);
+    }
+
+    private static final class XStat64Method extends BaseMethod {
+
+        private XStat64Method() {}
+
+        private static native int __xstat64(int version, String path, Stat64 stats);
+    }
+
+    private static final class StatMethod extends BaseMethod {
+
+        private StatMethod() {}
+
+        private static native int stat(String path, Stat64 stats);
+    }
+
+    /**
+     * Retrieves the actual number of bytes of disk storage used to store a specified file.
+     *
+     * @param path the path to the file
+     * @return an {@link OptionalLong} that contains the number of allocated bytes on disk for the file, or empty if the size is invalid
+     */
+    @Override
+    public OptionalLong allocatedSizeInBytes(Path path) {
+        assert Files.isRegularFile(path) : path;
+        final Stat64 stats = new Stat64();
+
+        final int rc;
+        if (method == Stat64Method.class) {
+            rc = Stat64Method.stat64(path.toString(), stats);
+        } else if (method == XStat64Method.class) {
+            rc = XStat64Method.__xstat64(1, path.toString(), stats);
+        } else {
+            assert method == StatMethod.class;
+            rc = StatMethod.stat(path.toString(), stats);
+        }
+        if (rc != 0) {
+            final int err = Native.getLastError();
+            logger.warn("error [{}] when executing native method stat(const char *path, struct stat *buf) for file [{}]", err, path);
+            return OptionalLong.empty();
+        }
+        if (logger.isTraceEnabled()) {
+            logger.trace("executing native method stat() returned {} for file [{}]", stats, path);
+        }
+        return OptionalLong.of(stats.st_blocks * ST_BLOCKS_UNIT);
+    }
+
+    /**
+     * Linux X86-64 for stat()
+     */
+    public static class Stat64 extends Structure {
+
+        public long st_dev;         // dev_t
+        public long st_ino;         // ino_t
+        public long st_nlink;       // nlink_t
+        public int st_mode;         // mode_t
+        public int st_uid;          // uid_t
+        public int st_gid;          // gid_t
+        public int pad0;            // int
```

Comment:
Oof bet it was fun to discover that this was needed...

Reply:
It came from a `struct` stat64 but now I can't find it anymore.